### PR TITLE
Add exception handling mechanism for process and service host.

### PR DESCRIPTION
**Pull request by HuanhuanSunMSFT**
1. Starting VS with poshtools installed will automatically start the service host process and establish the    channel.
2. If the host process is terminated, it won't be recreated until next IntelliSense is triggered.
3. If the channel is in a faulted state or closed, a new channel will be recreated.
4. A noticeable defect with this design is first Intellisense trigger when starting VS or recreating the process/service channel usually takes around ONE second.

@Microsoft/poshtools

### CODE REVIEW

**Comment by ASayre** [comment]
Why do we need to make the ConnectionManager, its contents and the Intellisense service static?  How will that impact our testability of these components?

**Comment by HuanhuanSunMSFT** [comment]
First, by doing this we won’t initialize the process until users triggers Intellisense. Especially when users intentionally end the Powershell process, we don’t want to re-create it immediately.
Second, we can put all the channel connection and exception logics in ConnectionManager by exposing a static PowershellIntelliSenseService.

As for testability, we actually make it easier to test the most important component OpenClientConnection.

File: PowerShellTools/ServiceManagement/ChannelFactoryMaker.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+namespace PowerShellTools.ServiceManagement
+{
+    internal static class ChannelFactoryMaker<ServiceType> where ServiceType : class
+    {
+        private static ChannelFactory<ServiceType> _channelFactory;
+        private static object _syncObject = new object();
+
+        public static string EndPointAddress { get; set; }
+
+        public static ChannelFactory<ServiceType> CreateChannelFactory()
```
Comment (ericdevg) [inline PowerShellTools/ServiceManagement/ChannelFactoryMaker.cs]: Looks like this factory method is requiring the endpoint address, why not taking it as parameter? Coz consumer of this factory method does not have to know they need initialize a static field before suing this static method....

File: PowerShellTools/ServiceManagement/ConnectionManager.cs [hunk 2]
```diff
@@ -7,49 +8,87 @@ namespace PowerShellTools.ServiceManagement
...
+                if (_powershellIntelliSenseService == null)
+                {
+                    OpenClientConnection();
+                }
+                return _powershellIntelliSenseService;
             }
         }
 
-        private void OpenClientConnection()
+        public static void OpenClientConnection()
```
Comment (ericdevg) [inline PowerShellTools/ServiceManagement/ConnectionManager.cs]: Does this need to be public?

File: PowerShellTools/ServiceManagement/ChannelFactoryMaker.cs (hunk 1 above)
Comment (HuanhuanSunMSFT) [inline PowerShellTools/ServiceManagement/ChannelFactoryMaker.cs]: Check out the latest commit I just made…

File: PowerShellTools/ServiceManagement/ConnectionManager.cs (hunk 2 above)
Comment (HuanhuanSunMSFT) [inline PowerShellTools/ServiceManagement/ConnectionManager.cs]: Will fix this

**Comment by ASayre** [comment]
After in-person convo about the state-ness of the static methods, Huan alleviated my concerns ☺

Until next adventure,

-Andre

**Comment by ASayre** [comment]
Looks good, after @EricMSFT gives his OK

**Pull request merged**